### PR TITLE
Enable UniversalPayload

### PR DESCRIPTION
**Pull request by td36**
Personal build before send out the formal patchs

### CODE REVIEW

File: UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h [hunk 1]
```diff
@@ -15,8 +15,8 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #pragma pack (1)
 
 typedef struct {
-  PLD_GENERIC_HEADER   Header;
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h]: This change should be done in UefiPayloadPkg:define BootManagerMenuFile HOB

File: UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h [hunk 2]
```diff
@@ -15,8 +15,8 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
-  GUID                 PassedGUID;
+  UNIVERSAL_PAYLOAD_GENERIC_HEADER   Header;
+  GUID                               PassedGUID;
```
Comment (niruiyu) [inline UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h:19]: FileName

File: UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h [hunk 3]
```diff
@@ -15,8 +15,8 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
 } BOOT_MANAGER_MENU_FILE;
```
Comment (niruiyu) [inline UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h:20]: UNIVERSAL_PAYLOAD_BOOT_MANAGER_MENU

File: UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c [hunk 4]
```diff
@@ -286,3 +288,39 @@ PlatformBootManagerUnableToBoot (
...
+  VOID                               *GuidHob;
+  BOOT_MANAGER_MENU_FILE             *BootManagerMenuFile;
+  Status = EFI_SUCCESS;
+  GuidHob = GetFirstGuidHob (&gUniversalPayloadBootManagerMenuFileGuid);
+  DEBUG((DEBUG_ERROR,"gUniversalPayloadBootManagerMenuFileGuid\n"));
+  //
+  // Find the buffer information and update PCDs
+  //
+  if (GuidHob != NULL) {
+    DEBUG((DEBUG_ERROR,"gUniversalPayloadBootManagerMenuFileGuid2\n"));
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:318]: Please check the revision and length, reference
MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c
line 1748

File: UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h (hunk 3 above)
Comment (niruiyu) [reply UefiPayloadPkg/Include/Guid/BootManagerMenuFile.h:20]: Also the header file name can be "BootManagerMenu.h"

File: UefiPayloadPkg/UefiPayloadPkg.dec [hunk 5]
```diff
@@ -30,7 +30,7 @@
   gBmpImageGuid                           = { 0x878AC2CC, 0x5343, 0x46F2, { 0xB5, 0x63, 0x51, 0xF8, 0x9D, 0xAF, 0x56, 0xBA } }
 
   ##include/Guid/BootManagerMenuFile.h
-  gUniversalPayloadBootManagerMenuFileGuid = { 0xdf939333, 0x42fc, ox4b2a, { 0xa5, ox9e, 0xbb, 0xae, 0x82, 0x81, 0xfe, 0xef }}
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/UefiPayloadPkg.dec]: This change should be done in UefiPayloadPkg:define BootManagerMenuFile HOB

File: UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c [hunk 6]
```diff
@@ -286,3 +288,39 @@ PlatformBootManagerUnableToBoot (
...
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  ) 
+{
+  EFI_STATUS                         Status;
+  UINTN                              size;
+  VOID                               *GuidHob;
+  BOOT_MANAGER_MENU_FILE             *BootManagerMenuFile;
+  Status = EFI_SUCCESS;
+  GuidHob = GetFirstGuidHob (&gUniversalPayloadBootManagerMenuFileGuid);
+  DEBUG((DEBUG_ERROR,"gUniversalPayloadBootManagerMenuFileGuid\n"));
```
Comment (niruiyu) [inline UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:313]: Remove this DEBUG()

File: UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c [hunk 7]
```diff
@@ -286,3 +288,39 @@ PlatformBootManagerUnableToBoot (
...
+  //
+  // Find the buffer information and update PCDs
+  //
+  if (GuidHob != NULL) {
+    DEBUG((DEBUG_ERROR,"gUniversalPayloadBootManagerMenuFileGuid2\n"));
+    BootManagerMenuFile = (BOOT_MANAGER_MENU_FILE *)GET_GUID_HOB_DATA (GuidHob);
+    size = sizeof(BootManagerMenuFile->PassedGUID);
+    Status = PcdSetPtrS (PcdBootManagerMenuFile, &size, &BootManagerMenuFile->PassedGUID);
```
Comment (niruiyu) [inline UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c:321]: 1. Check AutoGen.c/h to make sure the PcdSetPtrS() macro is expanded as expected.
2. size should be Size.
Did you run ECC to check coding style?

File: UefiPayloadPkg/UefiPayloadPkg.dsc [hunk 8]
```diff
@@ -363,6 +363,7 @@
 ################################################################################
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/UefiPayloadPkg.dsc:362]: I will add Cc list

File: UefiPayloadPkg/UefiPayloadPkg.dec [hunk 9]
```diff
@@ -72,3 +72,8 @@ gUefiPayloadPkgTokenSpaceGuid.PcdMemoryTypeEfiRuntimeServicesCode|0x80|UINT32|0x
 
 # Size of the region used by UEFI in permanent memory
 gUefiPayloadPkgTokenSpaceGuid.PcdSystemMemoryUefiRegionSize|0x02000000|UINT32|0x00000017
+
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/UefiPayloadPkg.dec:78]: Remove extra new lines

File: UefiPayloadPkg/UefiPayloadEntry/LoadDxeCore.c [hunk 10]
```diff
@@ -172,7 +175,11 @@ FvFindFile (
     //
     if (File->Type == FileType) {
       *FileHeader = File;
-      return EFI_SUCCESS;
+      DEBUG ((DEBUG_ERROR, "FileHeader->Name = %g\n", File->Name));
```
Comment (niruiyu) [inline UefiPayloadPkg/UefiPayloadEntry/LoadDxeCore.c:178]: Remove DEBUG()

File: UefiPayloadPkg/UefiPayloadEntry/LoadDxeCore.c [hunk 11]
```diff
@@ -279,11 +286,10 @@ LoadDxeCore (
   // Report DXE FV to DXE core
   //
   BuildFvHob ((EFI_PHYSICAL_ADDRESS) (UINTN) DxeCoreFv, DxeCoreFv->FvLength);
-
```
Comment (niruiyu) [inline UefiPayloadPkg/UefiPayloadEntry/LoadDxeCore.c:282]: Do not remove empty line.

**Review by niruiyu — CHANGES_REQUESTED** [review]

File: UefiPayloadPkg/UefiPayloadEntry/UniversalPayloadEntry.c [hunk 12]
```diff
@@ -265,6 +273,25 @@ BuildHobs (
     Hob.Raw = GET_NEXT_HOB (Hob);
   }
 
+  //
+  // Get DXE FV location
+  //
+  GuidHob = GetFirstGuidHob(&gUniversalPayloadExtraDataGuid);
+  ASSERT (GuidHob != NULL);
+  ExtraData = (UNIVERSAL_PAYLOAD_EXTRA_DATA *) GET_GUID_HOB_DATA (GuidHob);
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/UefiPayloadEntry/UniversalPayloadEntry.c:335]: Check the revision and length

File: UefiPayloadPkg/UefiPayloadEntry/UniversalPayloadEntry.c [hunk 13]
```diff
@@ -265,6 +273,25 @@ BuildHobs (
...
+  ASSERT (GuidHob != NULL);
+  ExtraData = (UNIVERSAL_PAYLOAD_EXTRA_DATA *) GET_GUID_HOB_DATA (GuidHob);
+  ASSERT (ExtraData->Count == 1);
+  ASSERT (AsciiStrCmp (ExtraData->Entry[0].Identifier, "uefi_fv") == 0);
+
+  *DxeFv = (EFI_FIRMWARE_VOLUME_HEADER *) (UINTN) ExtraData->Entry[0].Base;
+  ASSERT ((*DxeFv)->FvLength == ExtraData->Entry[0].Size);
+
+  //
+  // File DXE FV information to first fv hob in the hob list, which
```
Comment (niruiyu) [inline UefiPayloadPkg/UefiPayloadEntry/UniversalPayloadEntry.c:343]: Update DXE FV information to first FV hob in the hob list ...

File: UefiPayloadPkg/UefiPayloadEntry/UniversalPayloadEntry.c [hunk 14]
```diff
@@ -294,7 +324,13 @@ _ModuleEntryPoint (
...
 
   // Build HOB based on information from Bootloader
-  Status = BuildHobs (BootloaderParameter);
+  Status = BuildHobs (BootloaderParameter, &DxeFv);
+  ASSERT_EFI_ERROR (Status);
+
+  Status = UniversalLoadDxeCore (DxeFv, &DxeCoreEntryPoint);
+  ASSERT_EFI_ERROR (Status);
+
+  DEBUG ((DEBUG_INFO, "DxeCoreEntryPoint = 0x%lx\n", DxeCoreEntryPoint));
```
Comment (niruiyu) [inline UefiPayloadPkg/UefiPayloadEntry/UniversalPayloadEntry.c:388]: can you remove this DEBUG()?

**Review by niruiyu — CHANGES_REQUESTED** [review]

File: UefiPayloadPkg/Include/Library/DxeHobListLib.h [hunk 15]
```diff
@@ -0,0 +1,27 @@
+/** @file
```
Comment (LiuZhiguang001) [inline UefiPayloadPkg/Include/Library/DxeHobListLib.h:1]: should we change the name of DxeHobListLib?  Payload entry will also linke it. @niruiyu

**Review by niruiyu — APPROVED** [review]

File: UefiPayloadPkg/Include/Library/DxeHobListLib.h (hunk 15 above)
Comment (niruiyu) [reply UefiPayloadPkg/Include/Library/DxeHobListLib.h:1]: there are two HobList in UefiPayloadPkg: DxeHobLib and PayloadEntryHobLib. The latter one is linked to PayloadEntry.

**Pull request closed without merging**